File: py-sdks/src/pysdks/bcs/writer.py
"""BCS byte writer."""

from __future__ import annotations

from dataclasses import dataclass, field

from .uleb import encode_uleb128


@dataclass
class BCSWriter:
    _buf: bytearray = field(default_factory=bytearray)

    def write_u8(self, value: int) -> None:
        if not 0 <= value <= 0xFF:
            raise ValueError("u8 out of range")
        self._buf.append(value)

    def write_u16(self, value: int) -> None:
        self._write_int(value, 2)

    def write_u32(self, value: int) -> None:
        self._write_int(value, 4)

    def write_u64(self, value: int) -> None:
        self._write_int(value, 8)

    def write_bool(self, value: bool) -> None:
        self.write_u8(1 if value else 0)

    def write_bytes(self, value: bytes) -> None:
        self._buf.extend(value)

    def write_uleb128(self, value: int) -> None:
        self._buf.extend(encode_uleb128(value))

    def to_bytes(self) -> bytes:
        return bytes(self._buf)

    def _write_int(self, value: int, n: int) -> None:
        if value < 0:
            raise ValueError("negative integer not allowed")
        max_value = (1 << (8 * n)) - 1
        if value > max_value:
            raise ValueError(f"u{8*n} out of range")
        self._buf.extend(value.to_bytes(n, "little"))

File: dart-sdks/lib/src/bcs/writer.dart
import 'dart:typed_data';

import 'uleb.dart';

class BcsWriter {
  final BytesBuilder _builder = BytesBuilder(copy: false);

  void writeU8(int value) {
    _assertInRange(value, 0, 0xFF, 'u8');
    _builder.add([value]);
  }

  void writeU16(int value) => _writeInt(value, 2);

  void writeU32(int value) => _writeInt(value, 4);

  void writeU64(int value) => _writeInt(value, 8);

  void writeBool(bool value) {
    writeU8(value ? 1 : 0);
  }

  void writeBytes(List<int> value) {
    _builder.add(value);
  }

  void writeUleb128(int value) {
    _builder.add(encodeUleb128(value));
  }

  Uint8List toBytes() => _builder.toBytes();

  void _writeInt(int value, int len) {
    if (value < 0) {
      throw ArgumentError.value(value, 'value', 'negative integer not allowed');
    }
    final max = (BigInt.one << (8 * len)) - BigInt.one;
    if (BigInt.from(value) > max) {
      throw RangeError.value(value, 'value', 'u${8 * len} out of range');
    }

    final data = ByteData(len);
    switch (len) {
      case 2:
        data.setUint16(0, value, Endian.little);
        break;
      case 4:
        data.setUint32(0, value, Endian.little);
        break;
      case 8:
        data.setUint64(0, value, Endian.little);
        break;
      default:
        throw UnsupportedError('unsupported int width: $len');
    }
    _builder.add(data.buffer.asUint8List());
  }

  void _assertInRange(int value, int min, int max, String type) {
    if (value < min || value > max) {
      throw RangeError.value(value, 'value', '$type out of range');
    }
  }
}
